### PR TITLE
Cce analytic test exe prep, add option parsing for std::complex, add integer static cache range

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Adds the analytic test actions and makes some small alterations in preparation of the CCE analytic test executable.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Utilities/StaticCache.hpp [hunk 1]
```diff
@@ -24,6 +24,18 @@ struct CacheRange {
   using value_type = size_t;
 };
 
+/// \ingroup UtilitiesGroup
+/// Range of integral values for StaticCache indices.  The `Start` is inclusive
+/// and the `End` is exclusive.  The range must not be empty.
+template <typename IntegralType, IntegralType Start, IntegralType End>
```
Comment (kidder) [inline src/Utilities/StaticCache.hpp]: `template <auto Start, auto End>` and `static_assert(is_same_v<Start,End>)` and propagate that upward and downward

File: src/Options/StdParsers.hpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <complex>
```
Comment (kidder) [inline src/Options/StdParsers.hpp]: to reduce the potential for sucking in everything from the STL, please rename this file to `StdComplex` so we don't end up with a mega-file like StdHelpers

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 3]
```diff
@@ -9,31 +9,94 @@
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp:9]: add the word `actions` to the end of the commit message

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 4]
```diff
@@ -9,31 +9,94 @@
...
+struct H5WorldtubeBoundary;
+template <typename Metavariables>
+struct AnalyticWorldtubeBoundary;
+template <typename Metavariables>
+struct GhWorldtubeBoundary;
+/// \endcond
 namespace Actions {
+
+namespace detail {
+template <typename Initializer, typename ManagerTag, typename BoundaryCom>
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: `BoundaryCom` -> `BoundaryCommunicationTagsList`

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 5]
```diff
@@ -9,31 +9,94 @@
...
+struct InitializeWorldtubeBoundaryBase;
+
+template <typename WorldtubeComponent, template <typename> class Initializer,
+          typename ManagerTag, typename BoundaryCommunicationTags>
+struct InitializeWorldtubeBoundaryBase<Initializer<WorldtubeComponent>,
+                                       ManagerTag, BoundaryCommunicationTags> {
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: `BoundaryCommunicationTags` -> `BoundaryCommunicationTagsList`

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 6]
```diff
@@ -9,31 +9,94 @@
...
+  using initialization_tags = tmpl::list<ManagerTag>;
+  using initialization_tags_to_keep = tmpl::list<ManagerTag>;
+  using const_global_cache_tags = tmpl::list<Tags::LMax>;
+
+  using simple_tags = tmpl::list<::Tags::Variables<BoundaryCommunicationTags>>;
+
+  template <typename DbTags, typename... InboxTags, typename ArrayIndex,
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: `DbTags` -> `DataBoxTagsList`

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 7]
```diff
@@ -9,31 +9,94 @@
...
+          boundary_variables{
+              Spectral::Swsh::number_of_swsh_collocation_points(l_max)};
+
+      Initialization::mutate_assign<simple_tags>(make_not_null(&box),
+                                                 std::move(boundary_variables));
+      return std::make_tuple(std::move(box));
+    } else {
+      ERROR(MakeString{} << "Missing required boundary manager tag : "
+                         << db::tag_name<ManagerTag>);
+      return std::make_tuple(std::move(box));
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: don't need to return as `ERROR` should be marked `noreturn`

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 8]
```diff
@@ -48,109 +111,64 @@ namespace Actions {
...
-      tmpl::list<Tags::H5WorldtubeBoundaryDataManager>;
-  using initialization_tags_to_keep =
-      tmpl::list<Tags::H5WorldtubeBoundaryDataManager>;
+template <typename Metavariables>
+struct InitializeWorldtubeBoundary<H5WorldtubeBoundary<Metavariables>>
+    : public detail::InitializeWorldtubeBoundaryBase<
+          InitializeWorldtubeBoundary<H5WorldtubeBoundary<Metavariables>>,
+          Tags::H5WorldtubeBoundaryDataManager,
+          typename Metavariables::cce_boundary_communication_tags> {
+  using BaseType = detail::InitializeWorldtubeBoundaryBase<
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: BaseType -> base_type

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 9]
```diff
@@ -48,109 +111,64 @@ namespace Actions {
...
-  using initialization_tags = tmpl::list<Tags::GhInterfaceManager>;
-
-  using initialization_tags_to_keep = tmpl::list<Tags::GhInterfaceManager>;
+template <typename Metavariables>
+struct InitializeWorldtubeBoundary<GhWorldtubeBoundary<Metavariables>>
+    : public detail::InitializeWorldtubeBoundaryBase<
+          InitializeWorldtubeBoundary<GhWorldtubeBoundary<Metavariables>>,
+          Tags::GhInterfaceManager,
+          typename Metavariables::cce_boundary_communication_tags> {
+  using BaseType = detail::InitializeWorldtubeBoundaryBase<
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: BaseType -> base_type

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 10]
```diff
@@ -48,109 +111,64 @@ namespace Actions {
...
-            Spectral::Swsh::number_of_swsh_collocation_points(l_max)};
-
-    Initialization::mutate_assign<simple_tags>(make_not_null(&box),
-                                               std::move(boundary_variables));
-    return std::make_tuple(std::move(box));
-  }
+  using typename BaseType::initialization_tags;
+  using typename BaseType::initialization_tags_to_keep;
 };
+
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp:166]: remove

File: src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp [hunk 11]
```diff
@@ -64,7 +67,9 @@ class LinearizedBondiSachs;
  * calculations.
  */
 struct WorldtubeData : public PUP::able {
-  using creatable_classes = tmpl::list<BouncingBlackHole, LinearizedBondiSachs>;
+  using creatable_classes =
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp:70]: please test that all five of these can all be created from the factory

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 12]
```diff
@@ -9,31 +9,94 @@
...
+template <typename Metavariables>
+struct GhWorldtubeBoundary;
+/// \endcond
 namespace Actions {
+
+namespace detail {
+template <typename Initializer, typename ManagerTag, typename BoundaryCom>
+struct InitializeWorldtubeBoundaryBase;
+
+template <typename WorldtubeComponent, template <typename> class Initializer,
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: don't define a specialization, there is no use of `WorldtubeComponent`.  (i.e. remove lines 44-48)

File: src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp [hunk 13]
```diff
@@ -15,36 +15,21 @@
...
+ * \brief Generic base class for components that supply CCE worldtube boundary
+ * data. See class specializations for specific worldtube boundary components.
  */
-template <class Metavariables>
-struct H5WorldtubeBoundary {
-  using end_time_tag = Tags::EndTimeFromFile;
+template <typename WorldtubeComponent>
+struct WorldtubeComponentBase;
+
+template <class Metavariables, template <typename> class WorldtubeComponent>
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp]: just make the second template parameter to be `typename WorldtubeComponent`.  The template template parameter does not gain you anything (in fact it makes it less flexible)

File: src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp [hunk 14]
```diff
@@ -73,12 +58,49 @@ struct H5WorldtubeBoundary {
...
+ * `Metavariables` must contain:
+ * - the `enum` `Phase` with at least `Initialization` and `Evolve` phases.
+ * - a type alias `cce_boundary_communication_tags` for the set of tags to send
+ * from the worldtube to the evolution component. This will typically be
+ * `Cce::Tags::characteristic_worldtube_boundary_tags<Tags::BoundaryValue>`.
+ */
+template <class Metavariables>
+struct H5WorldtubeBoundary
+    : public WorldtubeComponentBase<H5WorldtubeBoundary<Metavariables>> {
+  using BaseType = WorldtubeComponentBase<H5WorldtubeBoundary<Metavariables>>;
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp]: BaseType -> base_type

File: src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp [hunk 15]
```diff
@@ -105,46 +127,17 @@ struct H5WorldtubeBoundary {
  * `Cce::Tags::characteristic_worldtube_boundary_tags<Tags::BoundaryValue>`.
  */
 template <class Metavariables>
-struct GhWorldtubeBoundary {
+struct GhWorldtubeBoundary
+    : public WorldtubeComponentBase<GhWorldtubeBoundary<Metavariables>> {
+  using BaseType = WorldtubeComponentBase<GhWorldtubeBoundary<Metavariables>>;
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp]: BaseType -> base_type

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 16]
```diff
@@ -406,8 +412,14 @@ struct SpecifiedEndTime : Tags::EndTime, db::SimpleTag {
   using option_tags = tmpl::list<OptionTags::EndTime>;
 
   static constexpr bool pass_metavariables = false;
-  static double create_from_options(const double end_time) noexcept {
-    return end_time;
+  static double create_from_options(
+      const std::optional<double> end_time) noexcept {
+    if(not static_cast<bool>(end_time)) {
+      ERROR(
+          "The start time must be explicitly specified for the tag "
```
Comment (kidder) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: start -> end?

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 17]
```diff
@@ -472,5 +481,21 @@ struct InitializeJ : db::SimpleTag {
   }
 };
 
+
```
Comment (kidder) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: remove

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 18]
```diff
@@ -170,5 +170,44 @@ struct InitializeWorldtubeBoundary<GhWorldtubeBoundary<Metavariables>>
...
+ * mechanism, so `Actions::SetupDataBox` must be present in the `Initialization`
+ * phase action list prior to this action.
+ */
+template <typename Metavariables>
+struct InitializeWorldtubeBoundary<AnalyticWorldtubeBoundary<Metavariables>>
+    : public detail::InitializeWorldtubeBoundaryBase<
+          InitializeWorldtubeBoundary<AnalyticWorldtubeBoundary<Metavariables>>,
+          Tags::AnalyticBoundaryDataManager,
+          typename Metavariables::cce_boundary_communication_tags> {
+  using BaseType = detail::InitializeWorldtubeBoundaryBase<
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: BaseType -> base_type

File: src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp [hunk 19]
```diff
@@ -123,6 +123,69 @@ struct BoundaryComputeAndSendToEvolution<H5WorldtubeBoundary<Metavariables>,
...
+                  boundary_variables) noexcept {
+            successfully_populated =
+                (*worldtube_data_manager)
+                    .populate_hypersurface_boundary_data(
+                        boundary_variables, time.substep_time().value());
+          });
+
+      if (not successfully_populated) {
+        ERROR("Insufficient boundary data to proceed, exiting early at time " +
+              std::to_string(time.substep_time().value()));
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/BoundaryComputeAndSendToEvolution.hpp]: can't you just stream the value?  (i.e.   `+ std::to_string(...)` -> `<< time.substep_time().value`

File: src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp [hunk 20]
```diff
@@ -101,6 +101,47 @@ struct H5WorldtubeBoundary
...
+ * `Metavariables` must contain:
+ * - the `enum` `Phase` with at least `Initialization` and `Evolve` phases.
+ * - a type alias `cce_boundary_communication_tags` for the set of tags to send
+ * from the worldtube to the evolution component. This will typically be
+ * `Cce::Tags::characteristic_worldtube_boundary_tags<Tags::BoundaryValue>`.
+ */
+template <class Metavariables>
+struct AnalyticWorldtubeBoundary
+    : public WorldtubeComponentBase<AnalyticWorldtubeBoundary<Metavariables>> {
+  using BaseType =
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp]: BaseType -> base_type

File: src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp [hunk 21]
```diff
@@ -101,6 +101,47 @@ struct H5WorldtubeBoundary
...
+ * `Evolve` phase.
+ *
+ * Uses const global tags:
+ * - `Tags::LMax`
+ * - `Tags::SpecifiedEndTime`
+ *
+ * `Metavariables` must contain:
+ * - the `enum` `Phase` with at least `Initialization` and `Evolve` phases.
+ * - a type alias `cce_boundary_communication_tags` for the set of tags to send
+ * from the worldtube to the evolution component. This will typically be
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp]: extra indentation in the bullets

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 22]
```diff
@@ -170,5 +170,44 @@ struct InitializeWorldtubeBoundary<GhWorldtubeBoundary<Metavariables>>
...
+ * \details Uses:
+ * - initialization tag
+ * `Cce::Tags::AnalyticBoundaryDataManager`,
+ * - const global cache tags `Tags::LMax`,
+ * `Tags::ExtractionRadius`.
+ *
+ * Databox changes:
+ * - Adds:
+ *   - `Tags::Variables<typename
+ * Metavariables::cce_boundary_communication_tags>`
```
Comment (kidder) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: extra indentation in bullet. (also did you render the documentation and look?)

**Review by kidder — CHANGES_REQUESTED** [review]
reviewed with @nilsdeppe

File: src/Utilities/StaticCache.hpp (hunk 1 above)
Comment (moxcodes) [reply src/Utilities/StaticCache.hpp]: woah, I missed that c++17 feature, that's slick :+1:

File: tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeLists.txt [hunk 23]
```diff
@@ -15,7 +15,7 @@ add_test_library(
   ${LIBRARY}
   "Elliptic/DiscontinuousGalerkin/"
   "${LIBRARY_SOURCES}"
-  "DataStructures;Domain;ErrorHandling"
+  "Boost::boost;DataStructures;Domain;ErrorHandling"
```
Comment (nilsvu) [inline tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeLists.txt]: AFAICT these tests don't use boost, so it shouldn't be linked here. A bunch of other libraries are missing (my fault), so perhaps adding them fixes the issue you had:

```
AnalyticSolutions;CoordinateMaps;DiscontinuousGalerkin;DomainCreators;DomainStructure;Elliptic;Initialization;LinearOperators;Parallel;ParallelDg;Spectral;Utilities
```

Even if this fixes the issue I don't understand why cmake would complain about boost here, because DataStructures actually links it in publicly... We should really convert our test libs to the new cmake structure 😅

**Review by nilsvu — CHANGES_REQUESTED** [review]
Not a full review. @kidder @nilsdeppe Do you want me to do a full review or is your joint review sufficient? :)

**Comment by kidder** [comment]
@nilsleiffischer, the joint review by @nilsdeppe and myself will be sufficient. Of course you are welcome to provide a review as well, but please let us know if you wish to do so.

**Comment by nilsvu** [comment]
No I'll leave it in your hands then :)

File: src/Options/StdParsers.hpp (hunk 2 above)
Comment (moxcodes) [reply src/Options/StdParsers.hpp]: I've squashed this filename change to make the code-change fixups less confusing

File: tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeLists.txt (hunk 23 above)
Comment (moxcodes) [reply tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeLists.txt]: I'm not sure precisely why, but either the boost include or the set you have suggested fix the problem, so I've added a fixup that switches it to your suggested set.

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp (hunk 10 above)
Comment (moxcodes) [reply src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp:166]: I've squashed this change in to the original commit to avoid making the fixups confusing or misordered.

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp (hunk 22 above)
Comment (moxcodes) [reply src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: oddly enough, it didn't seem to make a difference on my render. I've indented regardless.

**Review by moxcodes — COMMENTED** [review]
fixups posted, with a couple squashed to avoid fixup ordering irritations.
Cheers!

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp (hunk 22 above)
Comment (nilsdeppe) [reply src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: Sorry, I think Doxygen renders it correctly, I find it easier when I read the comments in a editor :P

File: tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeLists.txt [hunk 24]
```diff
@@ -15,7 +15,9 @@ add_test_library(
   ${LIBRARY}
   "Elliptic/DiscontinuousGalerkin/"
   "${LIBRARY_SOURCES}"
-  "Boost::boost;DataStructures;Domain;ErrorHandling"
+  "Elliptic;Utilities;Parallel;DomainStructure;CoordinateMaps;DomainCreators;\
```
Comment (kidder) [inline tests/Unit/Elliptic/DiscontinuousGalerkin/CMakeLists.txt]: we're not sure if @nilsleiffischer fixed this in the meantime, but please check

File: src/Utilities/StaticCache.hpp [hunk 25]
```diff
@@ -27,13 +27,14 @@ struct CacheRange {
 /// \ingroup UtilitiesGroup
 /// Range of integral values for StaticCache indices.  The `Start` is inclusive
 /// and the `End` is exclusive.  The range must not be empty.
-template <typename IntegralType, IntegralType Start, IntegralType End>
+template <auto Start, auto End>
```
Comment (kidder) [inline src/Utilities/StaticCache.hpp:18]: you should be able to now combine IntegralCacheRange and CacheRange

File: src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp [hunk 26]
```diff
@@ -14,7 +14,7 @@
 #include "Evolution/Systems/Cce/AnalyticSolutions/WorldtubeData.hpp"
 #include "Evolution/Systems/Cce/Tags.hpp"
 #include "Options/Options.hpp"
-#include "Options/StdParsers.hpp"
+#include "Options/StdComplex.hpp"
```
Comment (kidder) [inline src/Evolution/Systems/Cce/AnalyticSolutions/LinearizedBondiSachs.hpp:17]: this looks like it should belong in a different commit

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash everything addressing the below comments (looked at with @nilsdeppe )

**Comment by moxcodes** [comment]
Rebased on develop and fixups squashed in.
Cheers!

File: src/Utilities/StaticCache.hpp (hunk 25 above)
Comment (moxcodes) [reply src/Utilities/StaticCache.hpp:18]: this involved putting `_st` literal suffixes in a lot of places that otherwise passed `int` into the argument, so this commit changed a bit.

**Comment by kidder** [comment]
@moxcodes several clang-tidy errors to fix

**Comment by moxcodes** [comment]
build now fails in clang++-6.0 -- the error appears to be some very bad parse problem with the auto template parameters. I have a work-around, but maybe it's not worth putting all the casts in this code if we are just going to drop support for clang 6.0 in the next couple of days?

**Comment by nilsdeppe** [comment]
Yes, let's leave it until we've decided if we are dropping 6&7, if only 6 then we might need to test on 7, but I think it's pretty likely we will jump to 8.

**Comment by kidder** [comment]
@moxcodes do you need to do anything to this now that we've dropped clang6&7?

**Comment by moxcodes** [comment]
Yep, just needed a rebase to get the new CI workflow.

**Review by kidder — DISMISSED** [review]

**Comment by kidder** [comment]
@nilsdeppe @nilsleiffischer please look at

**Review by nilsdeppe — DISMISSED** [review]

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 27]
```diff
@@ -9,30 +9,89 @@
...
+  using initialization_tags_to_keep = tmpl::list<ManagerTag>;
+  using const_global_cache_tags = tmpl::list<Tags::LMax>;
+
+  using simple_tags =
+      tmpl::list<::Tags::Variables<BoundaryCommunicationTagsList>>;
+
+  template <
+      typename DataBoxTagsList, typename... InboxTags, typename ArrayIndex,
+      typename Metavariables, typename ActionList, typename ParallelComponent,
+      Requires<tmpl::list_contains_v<DataBoxTagsList, ManagerTag>> = nullptr>
```
Comment (nilsvu) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: I think this `Requires` is not needed because you have the `if constexpr` in the function below.

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 28]
```diff
@@ -9,30 +9,89 @@
...
+  static auto apply(db::DataBox<DataBoxTagsList>& box,
+                    const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    const Parallel::GlobalCache<Metavariables>& /*cache*/,
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    if constexpr (tmpl::list_contains_v<DataBoxTagsList, ManagerTag>) {
+      const size_t l_max = db::get<Tags::LMax>(box);
+      Variables<typename Metavariables::cce_boundary_communication_tags>
```
Comment (nilsvu) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: This should be `BoundaryCommunicationTagsList`

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 29]
```diff
@@ -9,30 +9,89 @@
...
+          boundary_variables{
+              Spectral::Swsh::number_of_swsh_collocation_points(l_max)};
```
Comment (nilsvu) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: You're missing an include for this

File: tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp [hunk 30]
```diff
@@ -113,6 +113,23 @@ struct SphericalSolutionWrapper : public SphericalSolution {
     }
   }
 
+  void test_serialize_and_deserialize(const size_t l_max,
+                                      const double time) noexcept {
+    const size_t size =
+        Spectral::Swsh::number_of_swsh_collocation_points(l_max);
```
Comment (nilsvu) [inline tests/Unit/Helpers/Evolution/Systems/Cce/AnalyticSolutions/AnalyticDataHelpers.hpp:120]: missing this include

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 31]
```diff
@@ -497,5 +506,20 @@ struct InitializeJ : db::SimpleTag {
...
+  using option_tags = tmpl::list<OptionTags::ExtractionRadius, OptionTags::LMax,
+                                 OptionTags::AnalyticSolution>;
+
+  static constexpr bool pass_metavariables = false;
+  static Cce::AnalyticBoundaryDataManager create_from_options(
+      const double extraction_radius, const size_t l_max,
+      const std::unique_ptr<Cce::Solutions::WorldtubeData>&
+          worldtube_data) noexcept {
+    return ::Cce::AnalyticBoundaryDataManager(l_max, extraction_radius,
+                                            worldtube_data->get_clone());
```
Comment (nilsvu) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: this formatting looks off. Perhaps just run `git-clang-format upstream/develop` and add the output as a separate commit?

**Review by nilsvu — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
@nilsleiffischer this is waiting on you to look at the fixups

**Comment by nilsvu** [comment]
Fixups LGTM, please rebase and squash.

Given this PR is pretty much ready to go I'm fine just merging it as-is, but it should be noted that at least the first commit `Add option parsing for std::complex<double>` should have been its own PR to make sure people involved with option parsing see the change. It's not apparent from the title of this PR that it touches code outside the CCE directories. Could you perhaps just make the title more descriptive, e.g. append `, add option parsing for std::complex, add integer static cache range`?

**Comment by moxcodes** [comment]
Squashed, rebased, and PR title adjusted.
Cheers!

**Review by nilsvu — DISMISSED** [review]
LGTM!

**Comment by kidder** [comment]
@moxcodes this is not compiling...

**Comment by moxcodes** [comment]
include problem fixed. This now passes CI.
Cheers!

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 32]
```diff
@@ -375,8 +375,14 @@ struct SpecifiedStartTime : Tags::StartTime, db::SimpleTag {
   using option_tags = tmpl::list<OptionTags::StartTime>;
 
   static constexpr bool pass_metavariables = false;
-  static double create_from_options(const double start_time) noexcept {
-    return start_time;
+  static double create_from_options(
+      const std::optional<double> start_time) noexcept {
+    if(not static_cast<bool>(start_time)) {
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: I'm happy for this to be changed in a followup PR: we've been preferring `start_time.has_value()` over the bool cast. I think `boost::optional` didn't have a `has_value()`, which is why we used to have the cast.

**Review by nilsdeppe — DISMISSED** [review]

**Review by nilsvu — DISMISSED** [review]

**Comment by moxcodes** [comment]
`has_value` change squashed in.
Cheers!

**Comment by nilsdeppe** [comment]
Awe merge conflict :(

**Comment by moxcodes** [comment]
Merge conflict solved.
Cheers!

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]
